README.md

```diff
@@ -1,34 +1,332 @@
-# Crate Name
+# vm-allocator crate
 
 ## Design
+vm-allocator is designed about resource and resource allocator being responsible
+for system resources allocation for virtual machine(VM).
 
-TODO: This section should have a high-level design of the crate.
+It provides different kinds of resources that might be used by VM, such as
+memory-maped I/O address space, port I/O address space, legacy IRQ numbers, MSI/MSI-X
+vectors, device instance id, etc. All these resources should be implemented with
+allocation and freeing mechanism in this crate, in order to make VMM easier to
+construct.
 
-Some questions that might help in writing this section:
-- What is the purpose of this crate?
-- What are the main components of the crate? How do they interact which each
-  other?
+Main components are listed below.
+- A `Resource` trait representing for all kinds of resources.
+```rust
+pub trait Resource {}
```

Comment:
Since SystemAllocator will be defined in the VMM, do we still need the traits to define `Resource`? What advantage do they bring in this particular case?

Reply:
This is not a "must have".

Reply:
I thought a bit more about this and I think there is one reason for having traits for `Resource`: with a `Resource` trait it means that vm-allocator does not need to depend on other crates (like virtio and vm-memory) for getting the actual types of the resources needed for allocating a resource. Does this assumption makes sense in the case of PCI and MSI? Ping @chao-p and @sboeuf.

I don't think that having a `ResourceAllocator` trait brings any value though. What do you think?

Reply:
Yes, using a trait looks fine to me since it might bring some flexibility for the future, if we were wanting to add another type of resource. I don't think that without the trait we have a dependency on virtio or vm-memory, since the vm-allocator is more like a fully independent component.

Reply:
What's the relationship between the vm-device and vm-allocator?
I think it should be that the vm-device crate depends on the vm-allocator crate.
And instead of define the Resource trait, we should actually define traits for MMIO address ranges, PIO ranges, IRQ descriptors.  For example, we may define:
enum InterrupConfig {
        LEGACY,
        MSI{addr_high: u32, addr_low: u32, data: u32},
}
trait InterruptSource {
       fn enable() -> Result<(), Error>;
       fn disable() -> Result<>;
       fn configure(config: InterruptConfig);
       fn set_mask(masked: bool);
       fn trigger(reason: u32);
       fn ack(reason: u32);
}

trait InterruptAllocator {
         fn allocate(count: u32) -> Result<Vec<impl InterruptSource>, Error>;
         fn free(Vec<impl InterruptSource>>);
}

For MMIO and PIO, it's similar with one special requirements for alignment because PCI bar enforces alignment. So the the allocate fn seems like:
trait MmioAddressAllocator {
    fn allocate(size, alignment, Option<min>, Option<max>);
}

Reply:
@jiangliu Or have 3 resource traits: an `Address` one that could be implemented for both MMIO and PIO, an `Interrupt` for MSI, INTX, etc, and an `Id` for allocating unique IDs.

Reply:
Sounds reasonable:)

Reply:
@jiangliu It seems you would like to use vm-device to manage MSI functions(InterruptSource) , not only manage the allocated vectors?